File: programmers/swift/60057.swift
// 문제 링크: https://programmers.co.kr/learn/courses/30/lessons/60057

func solution(_ s: String) -> Int {
    (1...(s.count / 2 + 1)).map { compress(s, $0).count }.min()!
}

func compress(_ string: String, _ length: Int) -> String {
    let string = Array(string)
    var result = ""

    var i = 0
    while i < string.count {
        var count = 0
        let substring = string[i..<min(i + length, string.count)]
        while i < string.count && string[i..<min(i + length, string.count)] == substring {
            count += 1
            i += length
        }

        result += count > 1 ? "\(count)\(String(substring))" : String(substring)
    }

    return result
}

File: programmers/python 3/60057.py
# 문제 링크: https://programmers.co.kr/learn/courses/30/lessons/60057

def solution(s):
    return min(map(
        lambda l: len(compress(s, l)),
        range(1, len(s) // 2 + 2)
    ))

def compress(s: str, l: int) -> str:
    result = ''

    i = 0
    while i < len(s):
        count = 0
        substr = s[i:min(i + l, len(s))]
        while i < len(s) and s[i:min(i + l, len(s))] == substr:
            count += 1
            i += l

        result += (str(count) if count > 1 else '') + substr

    return result